src/consensus/aft/async_execution_coordinator.h

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ds/thread_messaging.h"
+
+namespace aft
+{
+  enum class AsyncExecutionResult
+  {
+    PENDING,
+    COMPLETE
+  };
+
+  class AsyncExecutionCoordinator
+  {
+  public:
+    AsyncExecutionCoordinator(uint16_t thread_count) :
+      pending_cbs(0), run_sync(thread_count == 1)
+    {}
+
+    void increment_pending()
+    {
+      ++pending_cbs;
+    }
+
+    AsyncExecutionResult decrement_pending()
+    {
+      --pending_cbs;
+      return execution_status();
+    }
+
+    AsyncExecutionResult execution_status()
+    {
+      if (pending_cbs == 0) 
+      {
+        return AsyncExecutionResult::COMPLETE;
+      }
+      return AsyncExecutionResult::PENDING;
+    }
+
+    void start_next_execution_round(kv::Version last_idx)
+    {
+      is_first = true;
+      must_break = false;
+      execution_start_idx = last_idx;
+    }
+    
+    bool should_exec_next_append_entry(bool support_async_execution, uint64_t max_conflict_version)
+    {
+      if (!run_sync)
+      {
+        if (must_break)
+        {
+          return false;
+        }
+
+        if (execution_status() == AsyncExecutionResult::PENDING)
+        {
+          if (!support_async_execution && !is_first)
+          {
+            return false;
+          }
+
+          if (
+            support_async_execution &&
+            max_conflict_version >= execution_start_idx)
+          {
+            return false;
+          }
+        }
+      }
+
+      is_first = false;
+      must_break = !support_async_execution &&
+        (execution_status() == AsyncExecutionResult::PENDING);
+
+      return true;
+    }
+
+  private:
```

Comment:
We don't enforce this, but I think having the data members at the beginning of the class is better, and we do it in most places. Coming across them in the member functions and having to scroll down for them is not very intuitive.